[SR-2626] Fix String to Float/Double “snan” conversion

hartbit opened the pull request:
<!-- What's in this pull request? -->
As `strtof_l`/`strtod_l`/`swift_strtold_l` don’thandle `snan`, I added a manual check for that value in `_swift_stdlib_strtoX_clocale_impl`.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-2626](https://bugs.swift.org/browse/SR-2626).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

hartbit committed:
[SR-2626] Fix String to Float/Double “snan” conversion

As `strtof_l`/`strtod_l`/`swift_strtold_l` don’thandle `snan`, I added a manual check for that value in `_swift_stdlib_strtoX_clocale_impl`.

hartbit commented:
@stephentyrone @hughbe I implemented the fix in the non-Windows version of `_swift_stdlib_strtoX_clocale_impl`. I couldn't find information about `istringstream`'s support of `snan` as wasn't sure if it was necessary. On top of that, I wasn't happy about copy/pasting my modification in both functions. But not entirely sure what way you would have preferred to do the de-duplication.

xwu commented:
Per IEEE 754, this should work regardless of case and with an optional preceding sign. That is, "sNaN", "SNAN", "+snan", "-SnAn" should all work.

CodaFi commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -487,6 +487,12 @@ static const char *_swift_stdlib_strtoX_clocale_impl(
     const char * nptr, T* outResult, T huge,
     T (*posixImpl)(const char *, char **, locale_t)
 ) {
+  if (strcasecmp(nptr, "snan") == 0 || strcasecmp(nptr, "+snan") == 0 ||
```
Please use `strncasecmp` here.

hartbit replied:
Can you explain why? What if the string is "snangibberish"?

CodaFi replied:
Mostly paranoia.  You require the length of `nptr` to complete the return value anyways.  Computing it earlier and verifying it matches the length of the expected static strings is a more sound check.

xwu replied:
In that case, wouldn't it be nice to call `strncasecmp` only once?

```C++
  const std::size_t len = std::strlen(nptr);

  if ((len == 4 && strncasecmp(nptr, "snan", 4) == 0) ||
      (len == 5 && (nptr[0] == '+' || nptr[0] == '-') &&
       strncasecmp(nptr + 1, "snan", 4) == 0)) {
    // ...
  }
```

hartbit replied:
I don't think it would have better performance characteristics and is less readable. I'd still go for:

```
const std::size_t len = std::strlen(nptr);
if ((len == 4 && strncasecmp(nptr, "snan", len) == 0) ||
    (len == 5 && (strncasecmp(nptr, "+snan", len) || strncasecmp(nptr, "-snan", len))) {
  *outResult = std::numeric_limits<T>::signaling_NaN();
  return nptr + len;
}
 ||
      (len == 5 && (nptr[0] == '+' || nptr[0] == '-') &&
        strncasecmp(nptr + 1, "snan", 4) == 0)) {
    // ...
  }
```

xwu replied:
K, and this logic will need to either be copied to precede the `istringstream` or you'll need to extract it.

hartbit commented:
@xwu Thanks for the info! I corrected the code and moved the tests. Let me know if you have more comments.

xwu commented:
One more thing: Swift prints signaling NaN as "nan", which is fine. However, IEEE 754 says "if the conversion of a signaling NaN produces 'nan' or a sequence that is equivalent except for case, with an optional preceding sign, then the invalid operation exception should be signaled." We'll probably want logic in `swift_floatingPointToString` to do that.

hartbit commented:
I modified it to use `strncasecmp` and extracted the logic into a function to have it working on Windows. I can't test on Windows, but as `strncasecmp` doesn't exist there, I defined it to `_strnicmp`.

hartbit commented:
> One more thing: Swift prints signaling NaN as "nan", which is fine. However, IEEE 754 says "if the conversion of a signaling NaN produces 'nan' or a sequence that is equivalent except for case, with an optional preceding sign, then the invalid operation exception should be signaled." We'll probably want logic in swift_floatingPointToString to do that.

I'm not quite sure I understand the consequence of that IEEE 754 sentence? Does it just mean that we should return `snan` instead?

xwu commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -445,11 +445,27 @@ __mulodi4(di_int a, di_int b, int* overflow)
 #endif
 
 #if defined(__CYGWIN__) || defined(_WIN32)
+  #define strncasecmp _strnicmp
```
Cygwin has `strncasecmp`, as does MinGW. I think you'll want `#ifdef _MSC_VER` here, but I'm out of my depth here.

hartbit replied:
I'm out of my depth too. Lets wait for an official code review :)

stephentyrone replied:
This needs to switch `strncasecmp` as well, right?

hartbit replied:
`strcasecmp` you mean? Yes, I forgot to correct that line.

hartbit replied:
`strcasecmp` you mean? Yes, this is a mistake, I fixed in the latest push.

xwu commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -445,11 +445,27 @@ __mulodi4(di_int a, di_int b, int* overflow)
+  #define strncasecmp _strnicmp
+#endif
+
+static bool _swift_stdib_strissnan(const char *nptr) {
```
Perhaps, to help legibility, and since this isn't a shim for some libc function, we could give it a more readable name in the same vein as `swift_floatingPointToString`, such as `swift_stringIsSignalingNaN`?

hartbit replied:
Great suggestion. Done!

xwu commented:
> I'm not quite sure I understand the consequence of that IEEE 754 sentence? Does it just mean that we should return snan instead?

Returning `nan` is fine. However, it should actually signal. At least in my hands, signaling NaN doesn't seem to signal in Swift even when doing arithmetic, so this might be a discussion for another day.

stephentyrone commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -445,11 +445,27 @@ __mulodi4(di_int a, di_int b, int* overflow)
+    *outResult = std::numeric_limits<T>::signaling_NaN();
+    return nptr + std::strlen(nptr);
+  }
+  
```
There's no reason to use `strncasecmp` instead of `strcasecmp` when comparing to a literal; it makes your code more complex and doesn't add any safety (the literal necessarily has a trailing NUL, and `strcasecmp` will necessarily stop when it reaches it), and forces you to traverse the string twice. I would tend to write this function like so:
~~~~
if (nptr[0] == '+' || nptr[0] == '-')
  nptr++;
return (0 == strcasecmp(nptr, "snan"));
~~~~
(I would like it even more if the logic to strip a leading sign character were centralized somewhere, but this is acceptable).

hartbit replied:
Thanks for the suggestions. I applied the to the latest push!

stephentyrone commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -487,6 +504,11 @@ static const char *_swift_stdlib_strtoX_clocale_impl(
     T (*posixImpl)(const char *, char **, locale_t)
 ) {
+  if (swift_stringIsSignalingNaN(nptr)) {
+    *outResult = std::numeric_limits<T>::signaling_NaN();
```
In the long run we'll probably want to control the actual signaling NaN returned here so that the bit pattern of the significand is `b01xx ... xx`, but there's not a very tidy way to do that in C++ (which doesn't have anything like the Swift nan-with-payload initializer) without constructing from an explicit bit pattern. Can you please add the following TODO here to track this?
~~~~
// TODO: ensure that the returned sNaN bit pattern matches that of sNaNs produced by Swift.
~~~~

hartbit replied:
Done.

stephentyrone reviewed (approved):
LGTM. Thanks!

stephentyrone commented:
@swift-ci Please test.

hartbit commented:
@stephentyrone Why does the swift-ci comment mention commit 655e379? That was my first commit, but I have since forced push many more commits. The most recent one is 5109b12.

stephentyrone commented:
@hartbit I *think* that's just noise. give it a couple minutes, and I'll check back.

stephentyrone commented:
@hartbit it looks like 5109b12 was tested to me. Can you please confirm?

stephentyrone commented:
@hartbit Looks good, thanks again!
